# <center><font color='MediumSeaGreen'> Desafío Evaluado- Funciones y probabilidad básica</font><center>
### <center><b> Bárbara Pérez </b><center>

_________________

## <font color='MediumSeaGreen'>Desafío 1: Generación de funciones</font>

- Genere funciones para calcular la media y varianza de un vector. Debe cumplir con los siguientes requisitos:
    - Ambas funciones deben ingresar un argumento **x** correspondiente al vector.
    - Las funciones deben contener <span style='background:gainsboro'><font color='dimgray'>docstrings</font></span> con la documentación asociada a la variable y retornar un valor utilizando <span style='background:gainsboro'><font color='MediumSeaGreen'>returns</font></span>.
    - La función de la varianza se debe llamar a la función de la media

- Utilice las funciones para reportar la información sobre <span style='background:gainsboro'><font color='dimgray'>goles_favor</font></span>, <span style='background:gainsboro'><font color='dimgray'>goles_contra</font></span> y <span style='background:gainsboro'><font color='dimgray'>puntos</font></span>.

**(1 punto)**

In [1]:
##CREAR FUNCIONES

#Crear función de la media:
def media(vector):
    """
    media: retorna el promedio de una serie de numeros
        Parameters: 
             vector (int): serie numerica
        Return:
            float: promedio de los elementos de la serie
    """
    suma = 0 #Inicia el acumulador
    
    for i in vector:
        suma += i
        
    promedio = round((suma/len(vector)),2) #Calculo el promedio
    return promedio #Retorno el resultado


#Crear función de la varianza:
def varianza(vector):
    """
    varianza: retorna la varianza de una serie de numeros
    Parameters:
        vector (int): serie numerica
    Return:
        float: varianza de los elementos de la serie 
    """
    promedio = media(vector) #Obtengo el promedio del vector
    suma = 0 #Inicio del acumulador
    
    for i in vector: #Itera sobre el vector
        suma += (i-promedio)**2
        
    varianza = round((suma/(len(vector)-1)),2) #Obtengo la varianza
    return varianza #Retorno el resultado


#Crear función de la desviación estandar:
def desviacion_estandar(vector):
    """
    desviacion_estandar: retorna la desviacion estandar de una serie de numeros
    Parameters:
        vector (int): serie numerica
    Return:
        float: varianza de los elementos de la serie 
    
    """
    desv_estandar = round((varianza(vector)**(1/2)),2)
    return desv_estandar #Retorno el resultado

In [2]:
import pandas as pd
import numpy as np

In [3]:
#Importar archivo y crear DataFrame con su informacion
df = pd.read_csv('worldcup2014.csv')

In [4]:
#Variables del df
df.columns

Index(['team', 'continent', 'group', 'group_pos', 'cantidad_juegos',
       'juegos_ganados', 'juegos_empatados', 'juegos_perdidos', 'goles_favor',
       'goles_contra', 'goles_diferencia', 'puntos', 'clasificado'],
      dtype='object')

__________

**Utilice las funciones para reportar la información sobre goles_favor**

In [5]:
#Crear variable con datos del vector goles a favor
goles_favor = df['goles_favor']

In [6]:
#IMPRIMIR LOS RESULTADOS
print("Reporte la información sobre goles_favor:\n")

#Media del vector
print("-La media de 'goles_favor' es:", media(goles_favor))

#Varianza del vector
print("-La varianza de 'goles_favor' es:",varianza(goles_favor))

#Desviacion estandar del vector
print("-La desviación estandar de 'goles_favor' es:",desviacion_estandar(goles_favor))

Reporte la información sobre goles_favor:

-La media de 'goles_favor' es: 4.25
-La varianza de 'goles_favor' es: 5.35
-La desviación estandar de 'goles_favor' es: 2.31


_____

**Utilice las funciones para reportar la información sobre goles_contra.**

In [7]:
#Crear variable con datos del vector goles en contra 
goles_contra = df['goles_contra']

In [8]:
#IMPRIMIR LOS RESULTADOS
print("Reporte la información sobre goles_contra:\n")

#Media del vector
print("-La media de 'goles_contra' es:", media(goles_contra))

#Varianza del vector
print("-La varianza de 'goles_contra' es:",varianza(goles_contra))

#Desviacion estandar del vector
print("-La desviación estandar de 'goles_contra' es:",desviacion_estandar(goles_contra))

Reporte la información sobre goles_contra:

-La media de 'goles_contra' es: 4.25
-La varianza de 'goles_contra' es: 4.97
-La desviación estandar de 'goles_contra' es: 2.23


_________

**Utilice las funciones para reportar la información sobre puntos**

In [9]:
#Crear variable con datos del vector puntos
puntos = df['puntos']

In [10]:
#IMPRIMIR LOS RESULTADOS
print("Reporte la información sobre puntos:\n")

#Media del vector
print("-La media de 'puntos' es:", media(puntos))

#Varianza del vector
print("-La varianza de 'puntos' es:",varianza(puntos))

#Desviacion estandar del vector
print("-La desviación estandar de 'puntos' es:",desviacion_estandar(puntos))

Reporte la información sobre puntos:

-La media de 'puntos' es: 4.22
-La varianza de 'puntos' es: 8.31
-La desviación estandar de 'puntos' es: 2.88


___________

## <font color='MediumSeaGreen'>Desafío 2:</font>

Utilizando el método <span style='background:gainsboro'><font color='dimgray'>groupby</font></span> de la clase <span style='background:gainsboro'><font color='dimgray'>DataFrame</font></span>, en conjunto con la función <span style='background:gainsboro'><font color='dimgray'>.agg</font></span> de pandas, calcule la media, la varianza y desviación estándar de la cantidad de goles a favor, en contra y de la cantidad de puntos por continente.

- ¿En qué continente se observa una mayor cantidad de goles a favor?

- ¿En qué continente se observa una mayor cantidad de goles en contra?

- ¿En qué continente se observa una mayor cantidad de puntos en promedio?

**(2 puntos)**


In [11]:
#Definirlos en una variable
gol_favor = df.groupby(['continent'])['goles_favor'].agg([media, varianza, desviacion_estandar])
gol_contra = df.groupby(['continent'])['goles_contra'].agg([media, varianza, desviacion_estandar])
puntos_ = df.groupby(['continent'])['puntos'].agg([media, varianza, desviacion_estandar])

#Imprimir toda la información
print(f"GOLES A FAVOR\n\n {gol_favor}\n")
print("----------------------------------------------------")
print(f"\nGOLES EN CONTRA\n\n {gol_contra}\n")
print("----------------------------------------------------")
print(f"\nPUNTOS\n\n {puntos_}")


GOLES A FAVOR

               media  varianza  desviacion_estandar
continent                                         
africa         3.60      3.30                 1.82
asia           2.25      0.92                 0.96
europe         4.77      6.86                 2.62
northamerica   3.25      2.25                 1.50
southamerica   5.67      4.67                 2.16

----------------------------------------------------

GOLES EN CONTRA

               media  varianza  desviacion_estandar
continent                                         
africa         5.60      4.80                 2.19
asia           6.25      4.25                 2.06
europe         4.00      3.83                 1.96
northamerica   3.50     11.00                 3.32
southamerica   2.83      0.57                 0.75

----------------------------------------------------

PUNTOS

               media  varianza  desviacion_estandar
continent                                         
africa         2.40      3.30  

- **¿En qué continente se observa una mayor cantidad de goles a favor?**

Se observan mayor cantidad de goles a favor en el continente **SurAmericano**

- **¿En qué continente se observa una mayor cantidad de goles en contra?**

Se observa una mayor cantidad de goles en contra en el **continente Asiatico**

- **¿En qué continente se observa una mayor cantidad de puntos en promedio?**

Se observa una mayor cantidad de puntos en promedio en el **continente SurAmericano**

___________

## <font color='MediumSeaGreen'>Desafío 3: Simulaciones</font>

- Genere una función <span style='background:gainsboro'><font color='dimgray'>generate_pet</font></span> que devuelva de forma aleatoria un string <span style='background:gainsboro'><font color='FireBrick'>'perro'</font></span> o <span style='background:gainsboro'><font color='FireBrick'>'gato'</font></span> un número **n** de veces. Ejecútela un par de veces.

**Tip:** Puede utilizar la función <span style='background:gainsboro'><font color='dimgray'>np.random.choice</font></span> para retornar elementos al azar.

- Aplique la función <span style='background:gainsboro'><font color='dimgray'>generate_pet</font></span> para generar 20 muestras.

- ¿Cuál es la probabilidad de elegir un perro al azar? ¿Y un gato?

- Agregue <span style='background:gainsboro'><font color='dimgray'>np.random.seed(2)</font></span> al inicio del chunk. ¿Qué diferencia hay cuando se ejecuta la función varias veces luego de fijar la semilla?.

**(3 puntos)**

In [12]:
#Importar numpy
import numpy as np 

#Crear función generate_pet()
def generate_pet():
    """
        generate_pet: retorna un string aleatorio cuyos valores pueden ser 'perro' o 'gato'.
        Parameters: no tiene
        Return: (str): 'perro' o 'gato'        
    """
    
    return np.random.choice(['perro', 'gato']) #retorno el resultado

In [13]:
#Probar la función creada
generate_pet()

'perro'

In [14]:
#Crear función para crear lista mientras con gatos y perros de manera aleatorea
def obtiene_mascota(cantidad, semilla = 0):
    es_perro = 0
    es_gato = 0
    lista = []
    if semilla != 0:
        np.random.seed(semilla)
    for i in range(cantidad):
        mascota = generate_pet()
        lista.append(mascota)
        if mascota == 'perro':
            es_perro += 1
        else:
            es_gato += 1
            
    print(f"perros: {es_perro}, gatos: {es_gato}\n", lista)
    return es_perro, es_gato, lista 
    

In [15]:
#Aplicar la función generate_pet para generar 20 muestras.
es_perro, es_gato, lista = obtiene_mascota(20)

perros: 10, gatos: 10
 ['perro', 'gato', 'gato', 'perro', 'perro', 'gato', 'gato', 'perro', 'gato', 'gato', 'perro', 'perro', 'gato', 'perro', 'perro', 'gato', 'perro', 'perro', 'gato', 'gato']


**¿Cuál es la probabilidad de elegir un perro al azar? ¿Y un gato?**


In [16]:
#Crear variables para entregar porcentajes de la probabilidad de elegir un perro o un gato al azar
perro_azar = round((es_perro/len(lista)),2)
gato_azar = round((es_gato/len(lista)),2)

print(f"La probabilidad de elegir perro es de un {perro_azar*100} % ")
print(f"La probabilidad de elegir gato es de un {gato_azar*100} % ")

La probabilidad de elegir perro es de un 50.0 % 
La probabilidad de elegir gato es de un 50.0 % 


Según la lista creada anteriormente tiene un :

- 50% de probabilidades de elegir un perro al azar.

- 50% de probabilidades de elegir un gato al azar.

Sin embargo esta va a variar por el generador al azar "random.choice". Siendo lo más probable que cuando se vuelva a correr el chunk cambie su porcentaje.

_______

In [17]:
#Agreguar np.random.seed(2) al inicio del chunk.
def obtiene_mascota2(cantidad, semilla = 2):
    es_perro = 0
    es_gato = 0
    lista = []
    if semilla != 0:
        np.random.seed(semilla)
    for i in range(cantidad):
        mascota = generate_pet()
        lista.append(mascota)
        if mascota == 'perro':
            es_perro += 1
        else:
            es_gato += 1
            
    print(f"perros: {es_perro}, gatos: {es_gato}\n", lista)
    return es_perro, es_gato, lista 
    

In [18]:
#Aplicar la función generate_pet para generar 20 muestras.
es_perro, es_gato, lista = obtiene_mascota2(20)

perros: 8, gatos: 12
 ['perro', 'gato', 'gato', 'perro', 'perro', 'gato', 'perro', 'gato', 'perro', 'gato', 'perro', 'gato', 'gato', 'gato', 'gato', 'gato', 'gato', 'gato', 'perro', 'perro']


In [19]:
#Crear variables para entregar porcentajes de la probabilidad de elegir un perro o un gato al azar
perro_azar = round((es_perro/len(lista)),2)
gato_azar = round((es_gato/len(lista)),2)

print(f"La probabilidad de elegir perro es de un {perro_azar*100} % ")
print(f"La probabilidad de elegir gato es de un {gato_azar*100} % ")

La probabilidad de elegir perro es de un 40.0 % 
La probabilidad de elegir gato es de un 60.0 % 


**¿Qué diferencia hay cuando se ejecuta la función varias veces luego de fijar la semilla?.**

Cuando se fija la semilla, tambien se fija el resultado obtenido, por lo que el resultado no va a variar segun las veces que el usuario vuelva a correr el chunk siempre tendrá un resultado repetible. 

La semilla nos permitirá probar codigos con resultados repetibles esperados. 

En este caso se espera que se mantenga el siguiente resultado:
- La probabilidad de elegir perro es de un 40.0 % 

- La probabilidad de elegir gato es de un 60.0 % 

__________

## <font color='MediumSeaGreen'>Desafío 4: Función simuladora</font>

- Genere una función llamada <span style='background:gainsboro'><font color='dimgray'>simulate_pets_prob</font></span> que tome como argumento un número finito de simulaciones a generar.

- La función debe simular dos situaciones <span style='background:gainsboro'><font color='dimgray'>young_pet</font></span> y <span style='background:gainsboro'><font color='dimgray'>old_pet</font></span>, y contar la ocurrencia de los siguientes casos:

    - De las dos mascotas simuladas (young y old), contar las ocasiones donde por lo menos una de las mascotas sea un perro.
    
    - De las dos mascotas simuladas, contar las ocasiones donde <span style='background:gainsboro'><font color='dimgray'>old_pets</font></span> sea un perro.
    
    - De las dos mascotas simuladas, contar las ocasiones donde los dos sean perros.
    
- El método debe tener una semilla pseudoaleatoria de 1.

- El output de la función debe entregarse en términos de Probabilidad.

- De los tres escenarios, ¿Cuál es el menos probable? ¿Cuál es el más probable? ¿Por qué?

**(4 puntos)**

In [20]:
def simulate_pets_prob(cantidad, semilla = 1):
    
    es_perro = 0
    es_gato = 0
   
    
    young_pet = []
    
    if semilla == 1:
        np.random.seed(semilla)
    for i in range(cantidad):
        mascota = generate_pet()
        young_pet.append(mascota)
        
        if mascota == 'perro': 
            es_perro += 1
        else:
            es_gato += 1
          
    old_pets = []
    if semilla != 1:
        np.random.seed(semilla)
    for i in range(cantidad):
        mascota = generate_pet()
        old_pets.append(mascota)
        
        if mascota == 'perro': 
            es_perro += 1
        else:
            es_gato += 1 
    
    #contar las ocasiones donde por lo menos una de las mascotas sea un perro
    perro_1 = 0
    for mascota in young_pet or old_pets:    
        if mascota == 'perro':
            perro_1 += 1
        
    #contar las ocasiones donde old_pets sea un perro.
    perro_2 = 0
    for mascota in old_pets:    
        if mascota == 'perro':
            perro_2 += 1
    
    #contar las ocasiones donde los dos sean perros.
    perro_3 = 0
    for mascota in young_pet:    
        if mascota == 'perro':
            perro_3 += 1  
            
    perro_4 = perro_3 + perro_2
    
    
    return perro_1, perro_2, perro_4, young_pet, old_pets

In [21]:
perro_1, perro_2, perro_4, young_pet, old_pets = simulate_pets_prob(400)

In [22]:
escenario_1 = round((perro_1/len(young_pet + old_pets)),2)
print (f"La probabilidad que ocurra el escenario 1 es de: {escenario_1*100}%")

La probabilidad que ocurra el escenario 1 es de: 25.0%


In [23]:
escenario_2 = round((perro_2/len(old_pets)),2)
print (f"La probabilidad que ocurra el escenario 1 es de: {escenario_2*100}%")

La probabilidad que ocurra el escenario 1 es de: 54.0%


In [24]:
escenario_3 = round((perro_4/len(young_pet + old_pets)),2)
print (f"La probabilidad que ocurra el escenario 1 es de: {escenario_3*100}%")

La probabilidad que ocurra el escenario 1 es de: 52.0%


**De los tres escenarios:** 

**¿Cuál es el menos probable?**

El escenario menos probable es el escenario 1 (**al menos una de las mascotas sea un perro**). Esto se debe a que en la lista **young_pet** tiene una probabilidad de un 25% tener perro.

**¿Cuál es el más probable? ¿Por qué?**

El escenario más probable es el escenario 2 (**ocasiones donde old_pets sea un perro**), ya que en la lista **old_pet** tiene una probabilidad de 54%, siendo la probabilidad más altá entre los tres escenarios disponibles.

_________